### PR TITLE
Mount persistent volume to ActiveGate pod

### PR DESCRIPTION
**Pull request by andriisoldatenko**
## Description

This PR resolves https://dt-rnd.atlassian.net/browse/DAQ-1632

## How can this be tested?

**Test case _1_ (PVC disabled)**:
We set useEphemeral, which means (old behaviour, emptyDir)
```
apiVersion: dynatrace.com/v1beta3
kind: DynaKube
metadata:
  name: dynakube
  namespace: dynatrace
spec:
  apiUrl: https://hbc47170.dev.dynatracelabs.com/api
  oneAgent:
    applicationMonitoring:
  activeGate:
    useEphemeralVolume: true
    capabilities:
      - routing
```
- Checks:
- [x] pvc doesn't exists
- [x] we use empty dir as volume (as before)
```
k get pod dynakube-activegate-0 -o yaml | yq .spec.volumes
- emptyDir: {}
  name: ag-tmp-gateway
```

**Test case _2_ (default PVC)**:
We don't provide PVC, and don't set `useEphemeralVolume = true`
```
apiVersion: dynatrace.com/v1beta3
kind: DynaKube
metadata:
  name: dynakube
  namespace: dynatrace
spec:
  apiUrl: https://hbc47170.dev.dynatracelabs.com/api
  oneAgent:
    applicationMonitoring:
  activeGate:
    capabilities:
      - routing
```

Make sure default PVC has been created and used byt sts:
```
k get pvc ag-tmp-gateway-dynakube-activegate-0 -o yaml | yq .spec
accessModes:
  - ReadWriteOnce
resources:
  requests:
    storage: 1Gi
storageClassName: standard-rwo
volumeMode: Filesystem
volumeName: pvc-9d441346-f78c-4ca6-8dbe-e6862a60cf84

k get pod dynakube-activegate-0 -o yaml | yq .spec.volumes
- name: ag-tmp-gateway
  persistentVolumeClaim:
    claimName: ag-tmp-gateway-dynakube-activegate-0
```

**Test case _3_ (custom PVC)**:
```
apiVersion: dynatrace.com/v1beta3
kind: DynaKube
metadata:
  name: dynakube
  namespace: dynatrace
spec:
  apiUrl: https://hbc47170.dev.dynatracelabs.com/api
  oneAgent:
    applicationMonitoring:
      useCSIDriver: true
  activeGate:
    persistentVolumeClaim:
      accessModes:
        - ReadWriteOnce
      resources:
        requests:
          storage: 3Gi
    capabilities:
      - routing
```
same as previouse but your custom PVC.

```
k get pvc
NAME                                   STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   VOLUMEATTRIBUTESCLASS   AGE
ag-tmp-gateway-dynakube-activegate-0   Bound    pvc-c177436e-58c1-4643-b650-21d5b6a1b198   3Gi        RWO            standard-rwo   <unset>                 86s
```

```
k get pod dynakube-activegate-0 -o yaml | yq .spec.volumes
- name: ag-tmp-gateway
  persistentVolumeClaim:
    claimName: ag-tmp-gateway-dynakube-activegate-0
```

### CODE REVIEW

File: pkg/api/validation/dynakube/activegate.go [hunk 1]
```diff
@@ -17,6 +17,8 @@ Make sure you correctly specify the ActiveGate capabilities in your custom resou
 	errorDuplicateActiveGateCapability = `The DynaKube's specification tries to specify duplicate capabilities in the ActiveGate section, duplicate capability=%s.
 Make sure you don't duplicate an Activegate capability in your custom resource.
 `
+	errorActiveGateInvalidPVCConfiguration = `You provided a PVC and demanded to use an ephemeral volume at the same time for ActiveGate. These settings are mutual exclusive, please choose only one.`
```
[the remark this answers is not in the document]
Comment (andriisoldatenko) [reply pkg/api/validation/dynakube/activegate.go:20]: the message was copied from ticket, i'm okey to edit it, but prefer to double check with @StefanHauth

File: pkg/controllers/dynakube/activegate/internal/statefulset/statefulset.go [hunk 2]
```diff
@@ -119,6 +123,7 @@ func (statefulSetBuilder Builder) addTemplateSpec(sts *appsv1.StatefulSet) {
 		Affinity:           nodeAffinity(),
 		Tolerations:        statefulSetBuilder.capability.Properties().Tolerations,
 		SecurityContext: &corev1.PodSecurityContext{
+			FSGroup: address.Of(consts.DockerImageGroup),
```
[the remark this answers is not in the document]
Comment (andriisoldatenko) [reply pkg/controllers/dynakube/activegate/internal/statefulset/statefulset.go]: done.

File: pkg/controllers/dynakube/activegate/internal/statefulset/statefulset_test.go [hunk 3]
```diff
@@ -454,3 +454,85 @@ func TestSecurityContexts(t *testing.T) {
...
+			VolumeSource: corev1.VolumeSource{
+				EmptyDir: &corev1.EmptyDirVolumeSource{},
+			},
+		}
+		require.Contains(t, sts.Spec.Template.Spec.Volumes, expectedVolume)
+	})
+}
+
+func TestVolumeMounts(t *testing.T) {
+	t.Run("default OTLP ingest volume mount is presented in Container volumeMount list", func(t *testing.T) {
```
Comment (aorcholski) [inline pkg/controllers/dynakube/activegate/internal/statefulset/statefulset_test.go:483]:
```suggestion
	t.Run("volume mount is presented in Container volumeMount list", func(t *testing.T) {
```

it's always the same volume mount regardless of the current configuration. Maybe similar test for EphemeralVolume could be added?

File: pkg/api/validation/dynakube/activegate.go [hunk 4]
```diff
@@ -66,3 +68,17 @@ func missingActiveGateMemoryLimit(_ context.Context, _ *Validator, dk *dynakube.
 func memoryLimitSet(resources corev1.ResourceRequirements) bool {
 	return resources.Limits != nil && resources.Limits.Memory() != nil
 }
+
+func activeGateMutuallyExclusivePVCSettings(dk *dynakube.DynaKube) bool {
+	return dk.Spec.ActiveGate.UseEphemeralVolume && dk.Spec.ActiveGate.PersistentVolumeClaim != nil
+}
+
+func mutualExclusiveActiveGatePVsettings(_ context.Context, _ *Validator, dk *dynakube.DynaKube) string {
```
[the remark this answers is not in the document]
Comment (andriisoldatenko) [reply pkg/api/validation/dynakube/activegate.go]: fixed in 59a7aaea4

File: pkg/controllers/dynakube/activegate/internal/statefulset/statefulset_test.go (hunk 3 above)
Comment (andriisoldatenko) [reply pkg/controllers/dynakube/activegate/internal/statefulset/statefulset_test.go:504]: fixed in 98d85a5a4

File: hack/make/manifests/config.mk [hunk 5]
```diff
@@ -1,4 +1,4 @@
-CRD_OPTIONS ?= "crd:crdVersions=v1,maxDescLen=340,ignoreUnexportedFields=true"
+CRD_OPTIONS ?= "crd:crdVersions=v1,maxDescLen=315,ignoreUnexportedFields=true"
```
[the remark this answers is not in the document]
Comment (andriisoldatenko) [reply hack/make/manifests/config.mk:1]: i know i know :(

File: pkg/controllers/dynakube/activegate/internal/statefulset/statefulset.go [hunk 6]
```diff
@@ -112,22 +116,29 @@ func (statefulSetBuilder Builder) addUserAnnotations(sts *appsv1.StatefulSet) {
...
 func (statefulSetBuilder Builder) addTemplateSpec(sts *appsv1.StatefulSet) {
+	sc := corev1.PodSecurityContext{
+		SeccompProfile: &corev1.SeccompProfile{
+			Type: corev1.SeccompProfileTypeRuntimeDefault,
+		},
+	}
+
+	if !statefulSetBuilder.dynakube.Spec.ActiveGate.UseEphemeralVolume {
+		sc.FSGroup = address.Of(consts.DockerImageGroup)
+	}
```
[the remark this answers is not in the document]
Comment (andriisoldatenko) [reply pkg/controllers/dynakube/activegate/internal/statefulset/statefulset.go:127]: good idea, let me refactor.

File: pkg/api/v1beta3/dynakube/activegate/spec.go [hunk 7]
```diff
@@ -99,10 +102,15 @@ type Spec struct {
...
 
+	CapabilityProperties `json:",inline"`
+
 	// Activegate capabilities enabled (routing, kubernetes-monitoring, metrics-ingest, dynatrace-api)
 	Capabilities []CapabilityDisplayName `json:"capabilities,omitempty"`
 
 	enabledDependencies dependencies
+
+	// UseEphemeralVolume
+	UseEphemeralVolume bool `json:"useEphemeralVolume,omitempty"`
```
[the remark this answers is not in the document]
Comment (andriisoldatenko) [reply pkg/api/v1beta3/dynakube/activegate/spec.go:113]: we decided to stick to v1beta3, until we have final answers, and then we will introduce v1beta4, and conversion logic etc. cc @StefanHauth

File: pkg/controllers/dynakube/activegate/internal/statefulset/statefulset.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (andriisoldatenko) [reply pkg/controllers/dynakube/activegate/internal/statefulset/statefulset.go]: fixed in 210f035a4

File: pkg/api/v1beta3/dynakube/activegate/spec.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (StefanHauth) [reply pkg/api/v1beta3/dynakube/activegate/spec.go:113]: The AG will have a PVC _ONLY IF_ one of the features that require it are enabled. Those features will be

* OTLP ingest
* openSignal ingest
* maybe more to come

So if customers do not use those features, there won't be an PVC for them and everything continues to work as usual.

[the remark this answers is not in the document]
Comment (0sewa0) [reply pkg/api/v1beta3/dynakube/activegate/spec.go:113]: okok, will it be possible for them to provide a PVC, even if they don't use a feature that requires it?

**Review by aorcholski — APPROVED** [review]
LGTM

File: pkg/api/v1beta3/dynakube/activegate/spec.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (andriisoldatenko) [reply pkg/api/v1beta3/dynakube/activegate/spec.go:113]: @StefanHauth  could you please reply here?

[the remark this answers is not in the document]
Comment (StefanHauth) [reply pkg/api/v1beta3/dynakube/activegate/spec.go:113]: Ofc customers could provide one, but the PVC field should be ignored if no feature that requires a PVC is enabled.

[the remark this answers is not in the document]
Comment (0sewa0) [reply pkg/api/v1beta3/dynakube/activegate/spec.go:113]: seems a bit confusing to me, but fine 😅

File: config/crd/bases/dynatrace.com_edgeconnects.yaml [hunk 8]
```diff
@@ -339,7 +338,7 @@ spec:
                       Requests describes the minimum amount of compute resources required.
                       If Requests is omitted for a container, it defaults to Limits if that is explicitly specified,
                       otherwise to an implementation-defined value. Requests cannot exceed Limits.
-                      More info: https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/
+                      More info: https://kubernetes.
```
Comment (albertogdd) [inline config/crd/bases/dynatrace.com_edgeconnects.yaml:341]: Are these links cut intentionally?

Comment (andriisoldatenko) [reply config/crd/bases/dynatrace.com_edgeconnects.yaml:341]: it's due to this change https://github.com/Dynatrace/dynatrace-operator/pull/4084/files#diff-6118328086c863678a3dcf8134f32c75dc0558aed56214e0becfa6062161b9deR1
we cut entire description :(

Comment (albertogdd) [reply config/crd/bases/dynatrace.com_edgeconnects.yaml:341]: Sad.
I would remove then completely all these lines
```
More info: https://kubernetes.
```
just IMO

Comment (andriisoldatenko) [reply config/crd/bases/dynatrace.com_edgeconnects.yaml:341]: it's done automatically, i can't edit this file.

**Review by albertogdd — APPROVED** [review]

**Pull request merged**